core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -1753,7 +1792,20 @@ sealed trait ZIO[-R, +E, +A]
   final def retryUntilZIO[R1 <: R](
     f: E => URIO[R1, Boolean]
   )(implicit ev: CanFail[E], trace: Trace): ZIO[R1, E, A] =
-    self.catchAll(e => f(e).flatMap(b => if (b) Exit.fail(e) else ZIO.yieldNow *> retryUntilZIO(f)))
+    ZIO.suspendSucceed {
+      var result: Option[A] = None
+      ZIO
+        .whileLoop(result.isEmpty)(
+          self.foldZIO(
+            e => f(e).flatMap(b => if (b) Exit.fail(e) else Exit.unit),
+            { a =>
+              result = Some(a)
+              Exit.unit
+            }
+          )
+        )(ZIO.unitFn)
+        .as(result.get)
+    }
```

Comment:
OK this one is on me. For retries we have to yield on every iteration. This is because on single-threaded environments (i.e., either the host machine has 1 CPU or ScalaJS) if the effect is synchronous then we might end up hanging the application as the thread that is running the check never yields.

Let's revert the implementations of `retry`, `retryN` and `retryUntilZIO` as now that I'm looking at them more closely I don't think it makes much sense to change them.

Note that this will also resolve the issue with the failing CI